R/cran-compliance.R

```diff
@@ -0,0 +1,226 @@
+
+  # vendor directory will need to be ignored by git and R CMD build
+  if (!rlang::is_installed("usethis")) {
+    cli::cli_inform(
+      c(
+        "!" = "Add {.code ^src/rust/vendor$} to your {.file .Rbuildignore}",
+        "!" = "Add {.code ^src/rust/vendor$} to your {.file .gitignore}",
+        "i" = "Install {.pkg usethis} to have this done automatically."
+      )
+    )
+  } else {
+    # vendor will be big when expanded and should be ignored
+    usethis::use_build_ignore(
+      file.path("src", "rust", "vendor")
+    )
+
+    usethis::use_git_ignore(
+      file.path("src", "rust", "vendor")
+    )
+  }
+
+  invisible(NULL)
+}
+
+#' @export
+#' @name cran
+vendor_pkgs <- function(path = ".", quiet = FALSE, overwrite = NULL) {
+  local_quiet_cli(quiet)
+
+  # get path to rust folder
+  src_dir <- rprojroot::find_package_root_file(path, "src/rust")
+
+  # if `src/rust` does not exist error
+  if (!dir.exists(src_dir)) {
+    cli::cli_abort(
+      c("{.path src/rust} cannot be found", "i" = "Have you used {.fn use_extendr}?"),
+      class = "rextendr_error"
+    )
+  }
+
+  # if cargo.lock does not exist, cerate it using `cargo update`
+  cargo_lock_fp <- file.path(src_dir, "Cargo.lock")
+
+  if (!file.exists(cargo_lock_fp)) {
+    withr::with_dir(src_dir, {
+      update_res <- processx::run(
+        "cargo",
+        c(
+          "generate-lockfile",
+          "--manifest-path",
+          file.path(src_dir, "Cargo.toml")
+          ),
+        stderr_line_callback = function(x, proc) {
+          if (!grepl("To use vendored sources", x) && x != "") {
+            cli::cat_bullet(stringi::stri_trim_left(x))
+          }
+        }
+      )
+    })
+  }
+
+  if (update_res[["status"]] != 0) {
+    cli::cli_abort(
+      "{.file Cargo.lock} could not be creates with {.code cargo generate-lockfile}",
+      class = "rextendr_error"
+    )
+  }
+
+  # vendor crates
+  withr::with_dir(src_dir, {
+    vendor_res <- processx::run(
+      "cargo",
+      c(
+        "vendor",
+        "--locked",
+        "--manifest-path",
+        file.path(src_dir, "Cargo.toml")
+      ),
+      stderr_line_callback = function(x, proc) {
+        if (!grepl("To use vendored sources", x) && x != "") {
+          cli::cat_bullet(stringi::stri_trim_left(x))
+        }
+      }
+    )
+  })
+
+  if (vendor_res[["status"]] != 0) {
+    cli::cli_abort(
+      "{.code cargo vendor} was not executed successfully",
+      class = "rextendr_error"
+    )
+  }
+
+  # capture vendor-config.toml content
+  config_toml <- stringi::stri_split(vendor_res$stdout, coll = "\n")[[1]]
+
+  # always write to file as cargo vendor catches things like patch.crates-io
+  # and provides the appropriate configuration.
+  brio::write_lines(config_toml, file.path(src_dir, "vendor-config.toml"))
+  cli::cli_alert_info("Writing {.file src/rust/vendor-config.toml}")
+
+  # compress to vendor.tar.xz
+  compress_res <- withr::with_dir(src_dir, {
+    processx::run(
+      "tar", c(
```

Comment:
Does this work on windows? Is it calling some bundled with `Rtools` `tar`?

Reply:
I believe it does, I'm not sure what mechanism it is made available. I've added another expectation to the use-cran-defaults test to check that the tarred file gets created. If this passes on windows CI, then I'm comfortable saying yes